src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,459 @@ namespace System
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElements);
+            int length = ClampEnd(span, start, trimElements);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElements);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a set of elements specified
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        /// <remarks>If <paramref name="trimElements"/> is empty, the span is returned unaltered.</remarks>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElements);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits  all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElements">The span which contains the set of elements to remove.</param>
+        private static int ClampStart<T>(in ReadOnlySpan<T> span, in ReadOnlySpan<T> trimElements)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty)
+                return 0;
+
+            switch (trimElements.Length)
+            {
+                case 0:
+                    return 0;
+                case 1:
+                    return ClampStart(span, trimElements[0]);
+            }
+
+            int start = 0;
+
+            for (; start < span.Length; start++)
+            {
+                for (int i = 0; i < trimElements.Length; i++)
+                {
+                    if (trimElements[i] == null)
+                    {
+                        if (span[start] != null)
+                            goto Next;
+                    }
+                    else if (span[start] != null && !span[start].Equals(trimElements[i]))
+                    {
+                        goto Next;
+                    }
+                }
+                break;
+            Next:
```

Comment:
Rather than duplicating the above logic and also using gotos, could the inner loop just be extracted to a helper?

Reply:
Done. Since it's called in a (outer) loop I used aggressive inlining. Let me know if I shouldn't.